### PR TITLE
feat: add TVC over time data

### PR DESCRIPTION
**Pull request by frimoldi**
- List of protocols are being fetched from a local `.csv` file.
- Use DefiLlama API to get protocols' TVLs.
- TVC is being calculated and stored once per ~6400 blocks (~1 day).

### CODE REVIEW

File: indexer.py [hunk 1]
```diff
@@ -18,12 +20,15 @@
...
+    StatsTVC
 )
 from utils import get_event_logs_in_range, time_delta_apy
 
 YEAR = Decimal(timedelta(days=365).total_seconds())
 getcontext().prec = 78
-logging.basicConfig(filename="output.log", level=logging.INFO)
+
+logging.basicConfig(level=logging.INFO)
+# logging.basicConfig(filename="output.log", level=logging.INFO)
```
Comment (Evert0x) [inline indexer.py:31]: I think we can undo these changes

File: indexer.py [hunk 2]
```diff
@@ -45,10 +50,12 @@ def __init__(self, blocks_per_call=None):
             settings.SHERLOCK_PROTOCOL_MANAGER_WSS.events.ProtocolRemoved: self.ProtocolRemoved.new,
             settings.SHERLOCK_PROTOCOL_MANAGER_WSS.events.ProtocolRemovedByArb: self.ProtocolRemoved.new,
         }
-
-        # 268 blocks is roughly every hour on current Ethereum mainnet
```
Comment (Evert0x) [inline indexer.py:49]: Comment is removed

File: indexer.py [hunk 3]
```diff
@@ -82,6 +89,45 @@ def calc_tvl(self, session, indx, block):
...
+        protocols = {}
+        with open("./meta/protocols.csv", newline="") as csv_file:
+            csv_reader = csv.DictReader(csv_file)
+            for row in csv_reader:
+                if "id" in row:
+                    protocols[row["id"]] = row["defi_llama_slug"]
+
+        timestamp = settings.WEB3_WSS.eth.get_block(block)["timestamp"]
+        accumulated_tvc_for_block = 0
+        for protocol_bytes, defi_llama_slug in protocols.items():
```
Comment (Evert0x) [inline indexer.py:103]: This can be done in a single loop.

File: indexer.py [hunk 4]
```diff
@@ -82,6 +89,45 @@ def calc_tvl(self, session, indx, block):
 
         StatsTVL.insert(session, block, timestamp, tvl)
 
+    def calc_tvc(self, session, indx, block):
+        # read list of protocols from local csv file
+        protocols = {}
+        with open("./meta/protocols.csv", newline="") as csv_file:
```
Comment (Evert0x) [inline indexer.py:95]: Any changes to the `protocols.csv` can cause issues during runtime.
Maybe we can read from the csv in settings.py and parse to dictionary and use the in memory object here.

File: indexer.py [hunk 5]
```diff
@@ -82,6 +89,45 @@ def calc_tvl(self, session, indx, block):
...
+
+            protocol_coverages = ProtocolCoverage.get_protocol_coverages(session, protocol.id)
+
+            if not protocol_coverages:
+                continue
+
+            protocol_coverage = protocol_coverages[0]
+
+            # fetch protocol's TVL from DefiLlama
+            response = requests.get("https://api.llama.fi/protocol/" + defi_llama_slug)
```
Comment (Evert0x) [inline indexer.py:118]: Surround with try/catch and log any errors. Could be that the api service is offline, in this case we should just skip adding an entry.

File: indexer.py (hunk 4 above)
Comment (frimoldi) [reply indexer.py:98]: yeah, good call

**Review by rcstanciu — APPROVED** [review]
Looks good to me

File: indexer.py [hunk 6]
```diff
@@ -45,10 +49,13 @@ def __init__(self, blocks_per_call=None):
             settings.SHERLOCK_PROTOCOL_MANAGER_WSS.events.ProtocolRemoved: self.ProtocolRemoved.new,
             settings.SHERLOCK_PROTOCOL_MANAGER_WSS.events.ProtocolRemovedByArb: self.ProtocolRemoved.new,
         }
-
-        # 268 blocks is roughly every hour on current Ethereum mainnet
-        self.intervals = {self.calc_tvl: settings.INDEXER_STATS_BLOCKS_PER_CALL,
-                          self.calc_factors: 1, self.reset_apy_calc: 268}
+        self.intervals = {
+            self.calc_tvl: settings.INDEXER_STATS_BLOCKS_PER_CALL,
+            self.calc_tvc: 5,
```
Comment (Evert0x) [inline indexer.py:54]: Are we going to calculate TVC every 5 blocks? That seems a bit too much

File: indexer.py [hunk 7]
```diff
@@ -1,6 +1,8 @@
 import logging
+from msilib.schema import Error
```
Comment (Evert0x) [inline indexer.py:2]: What are we using this import for?

File: indexer.py [hunk 8]
```diff
@@ -82,6 +89,44 @@ def calc_tvl(self, session, indx, block):
...
+                for tvl_data_point in reversed(tvl_historical_data):
+                    if tvl_data_point["date"] < int(timestamp):
+                        # if protocol's TVL < coverage_amount => TVC = TVL, otherwise TVC = coverage_amount
+                        tvc = min(tvl_data_point["totalLiquidityUSD"] * 1000000, protocol_coverage.coverage_amount)
+                        accumulated_tvc_for_block += int(tvc)
+                        break
+
+            StatsTVC.insert(session, block, datetime.fromtimestamp(timestamp), accumulated_tvc_for_block)
+        except Exception as e:
+            logging.exception("Encountered exception %s" % e)
```
Comment (Evert0x) [inline indexer.py:128]: Maybe something more specific, like "TVC calculation encountered exception"

**Review by Evert0x — APPROVED** [review]
